net/core/filter.c

```diff
@@ -4792,6 +4792,259 @@ static const struct bpf_func_proto bpf_lwt_seg6_adjust_srh_proto = {
+	.func		= bpf_lua_pushinteger,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushlightuserdata, void *, ptr)
+{
+	lua_pushlightuserdata(L, ptr);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushlightuserdata_proto = {
+	.func		= bpf_lua_pushlightuserdata,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_2(bpf_lua_pushlstring, const char *, str, size_t, len)
+{
+	lua_pushlstring(L, str, len);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushlstring_proto = {
+	.func		= bpf_lua_pushlstring,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+	.arg2_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushmap, struct bpf_map *, map)
+{
+	lua_pushlightuserdata(L, map);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushmap_proto = {
+	.func		= bpf_lua_pushmap,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_CONST_MAP_PTR,
+};
+
+
+BPF_LUA_CALL_0(bpf_lua_pushskb)
+{
+	lua_pushlightuserdata(L, xdp_lua_get_skb());
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushskb_proto = {
+	.func		= bpf_lua_pushskb,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushstring, const char *, str)
+{
+	lua_pushstring(L, str);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushstring_proto = {
+	.func		= bpf_lua_pushstring,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_toboolean, int, index)
+{
+	return lua_toboolean(L, index);
+}
+
+static const struct bpf_func_proto bpf_lua_toboolean_proto = {
+	.func		= bpf_lua_toboolean,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_tointeger, int, index)
+{
+	return lua_tointeger(L, index);
+}
+
+static const struct bpf_func_proto bpf_lua_tointeger_proto = {
+	.func		= bpf_lua_tointeger,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_2(bpf_lua_newpacket, struct xdp_buff *, xdp, int, offset) {
+	if (offset + xdp->data < xdp->data_end) {
+		return lunpack_newpacket(L, xdp->data + offset,
+				xdp->data_end - xdp->data - offset);
+	}
+
+	return -EINVAL;
+}
+
+static const struct bpf_func_proto bpf_lua_newpacket_proto = {
+	.func		= bpf_lua_newpacket,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_PTR_TO_CTX,
+	.arg2_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_3(bpf_lua_tostring, char *, str, u32, size, int, index) {
+	if (lua_isstring(L, index)) {
+		const char *s;
+		size_t bytes;
+
+		s = lua_tolstring(L, index, &bytes);
+		if (bytes + 1 > size)
```

Comment:
It looks like the `+1` is already computed in bytes. The wording in Lua spec[1] doesn't make this explicit. Could you double check this?

```
lua_tolstring returns a fully aligned pointer to a string inside the Lua state.
This string always has a zero ('\0') after its last character (as in C), but can contain other zeros in its body.
```

Reply:
@mpolitzer apparently Lunatik uses `strlen`: https://github.com/luainkernel/lunatik/blob/ad05be89cd2674cdcef326f47c453cefce0eb580/lua/lstring.c#L233